azurerm/resource_arm_application_gateway.go

```diff
@@ -1340,331 +1339,423 @@ func flattenApplicationGatewayWafConfig(waf *network.ApplicationGatewayWebApplic
+				settings["timeout"] = int(*props.Timeout)
+				settings["unhealthy_threshold"] = int(*props.UnhealthyThreshold)
+			}
+
+			result = append(result, settings)
+		}
 	}
 
 	return result
 }
 
 func flattenApplicationGatewayRequestRoutingRules(rules *[]network.ApplicationGatewayRequestRoutingRule) ([]interface{}, error) {
-	result := make([]interface{}, 0, len(*rules))
-
-	for _, config := range *rules {
-		httpListenerName := strings.Split(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID, "/")[len(strings.Split(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID, "/"))-1]
-		listener := map[string]interface{}{
-			"id":                 *config.ID,
-			"name":               *config.Name,
-			"rule_type":          string(config.ApplicationGatewayRequestRoutingRulePropertiesFormat.RuleType),
-			"http_listener_id":   *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.HTTPListener.ID,
-			"http_listener_name": httpListenerName,
-		}
+	result := make([]interface{}, 0)
 
-		if config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool != nil {
-			backendAddressPoolName := strings.Split(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool.ID, "/")[len(strings.Split(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool.ID, "/"))-1]
-			listener["backend_address_pool_name"] = backendAddressPoolName
-			listener["backend_address_pool_id"] = *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendAddressPool.ID
-		}
+	if rules != nil {
+		for _, config := range *rules {
+			listener := map[string]interface{}{
+				"id":   *config.ID,
+				"name": *config.Name,
+			}
 
-		if config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendHTTPSettings != nil {
-			backenndHTTPSettingsName := strings.Split(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendHTTPSettings.ID, "/")[len(strings.Split(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendHTTPSettings.ID, "/"))-1]
-			listener["backend_http_settings_name"] = backenndHTTPSettingsName
-			listener["backend_http_settings_id"] = *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.BackendHTTPSettings.ID
-		}
+			if props := config.ApplicationGatewayRequestRoutingRulePropertiesFormat; props != nil {
+				listener["rule_type"] = string(props.RuleType)
 
-		if config.ApplicationGatewayRequestRoutingRulePropertiesFormat.URLPathMap != nil {
-			urlPathMapName := strings.Split(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.URLPathMap.ID, "/")[len(strings.Split(*config.ApplicationGatewayRequestRoutingRulePropertiesFormat.URLPathMap.ID, "/"))-1]
-			listener["url_path_map_name"] = urlPathMapName
-			listener["url_path_map_id"] = *config.ApplicationGatewayRequestRoutingRulePropertiesFormat.URLPathMap.ID
-		}
+				// TODO: can we switch these out for the Resource ID Parsers?
```

Comment:
The same format is all throughout this PR so I think it would be a good idea to make this a method either now or later

Reply:
yep - I've put these as a TODO which I'll pick up next to investigate if we can switch to using the `parseAzureResourceId` method which should solve this; if not we'll pull this out and refactor it into a separate method